### ISSUE TITLE
What's your end goal?

### ISSUE THREAD

**Issue by ilikenwf** [MatthewTingum/16---17-Cruze-Malibu-Infotainment-Hacking#1]
Hacking/rooting/modding?

I have a 2018 mylink...haven't done anything but remove the modem because it crepped me out...I do have a pcap somewhere of the packets exchanged when it checked for software updates, though...if that would help.

**Comment by MatthewTingum** [comment]
Ultimately I'd like to get a remote root shell. It's mostly just a boredom project, doing it for fun and to hopefully pick up a few things along the way.

That said, I could think of a couple uses. It could run some custom software to monitor and interact with GMLAN. Sure there are other tools to do this, but a built in solution would be convenient. I've got a couple of these radios from another project. My "replacement" unit won't work in my car because of anti-theft (anti-repair). This could potentially be bypassed.

I found a way to enter "engineering mode" which is different from the "development mode" I've seen on a few forums. This allows you to connect to an access point and check for FOTA ("Firmware Over the Air") updates. I logged some of this traffic, but it's https so I can't see much. Here is the update server it connects to: https://vtmpub.oboservices.mobi

If I can log traffic from within the radio, might be able to figure out how to make a fake update server (although the cert might prevent this). Alternatively reversing the updater might provide some clues.

dev / engr codes:
20130821 - untested
20130822 - dev mode (seen on forums)
1029475638185 - engineering mode

Linux support for QNX6 filesystems is read only so modifying the filesystem is a bit tricky. I've been editing the eMMC image and reflashing the entire thing every time I make a change. Trying to get it to boot a shell script off a thumb drive so I can run tests without reflashing every time. For whatever reason, this doesn't appear to be working right now.

I wouldn't mind taking a look at that pcap.

**Comment by ilikenwf** [comment]
I think mine is an IO6 unit, and I believe it's a Bosch unit based on Linux, not QNX...

**Comment by ilikenwf** [comment]
Also, I'm not entirely sure if useful or not but I found some qcm confidential stuff and some SDK's on this developers github... seems they dev for QNX. There's a lot more in there:

https://github.com/runmousefly/Work/blob/7d99feea7de35b8fc7a89fdc3ba4b14249422f07/MctCoreServices/src/com/mct/carmodels/RZC_GMSeriesProtocol.java

**Comment by MatthewTingum** [comment]
Hmmm, might have to pick up one of the newer units when I'm done with this one. Been looking for projects like this to dip my toes in the hardware side of things.

Good find! I'll have to dig in a bit. I've actually been looking for some of those values.

Still haven't gotten remote root access on mine. It should be trivial, but I'm more focused on other things for the time being.

**Comment by ilikenwf** [comment]
I think mine uses an i.MX6 CPU, also...there are some threads around the web about it.

The pcap I have is mostly over ssl and so only some hostnames are visible.

**Comment by MatthewTingum** [comment]
Gotcha, I'll look into it. Saw a datasheet but had more luck with intuition on the actual board.

vtmpub.oboservices.mobi renewed last week or something like that. Was hoping some administrative flaw would have prevented this. Not going to have any luck pushing down updates on that end.

Might be possible to create a local faked server, but I can just flash the whole thing, so that's not a concern for now.

My biggest hangups right now are:
IFS - cannot be edited. IDK how to properly repack them.
QNX FS - these drives need to be edited manually in a hex editor

**Comment by ilikenwf** [comment]
Yeah, same here...if you search github code for the oboservices.mobi domain you'll find some API references...

All that said, I know that for the IO6 and others that use Linux have the GPL code released on Bosch's open source site. It shows some of their configuration details and versions, which may be enough to exploit our way to root someday.

https://oss.bosch-cm.com/gm.html

**Comment by ilikenwf** [comment]
Also, this may interest you: 
https://github.com/search?q="Bosch+Car+Multimedia"&type=Code

**Comment by ilikenwf** [comment]
So looking, it appears that this system is 3 parts...the LG LTE/Nav/etc box, the Sirius Box, and the HMI. My HMI is version 2.5, and may be QNX based, although the license screens on the display show that it's Linux, maybe, or one of the three is.

The LG box is the one with an i.MX chip, the HMI is probably very similar if not the same as yours...I did attach a chip clip to the Winbond flash chip on my HMI, but have yet to investigate that...I realize the system ram is the soldered solid state chips, but...just some info.

I have pictures and stuff if you need them.

**Comment by ilikenwf** [comment]
https://gm-techlink.com/wp-content/uploads/2015/10/GM_TechLink_19_October_2015.pdf

**Comment by MatthewTingum** [comment]
Hey, thanks for the info! Haven't worked on this in a bit, but I'm never opposed to more knowledge. I'd love to see the pics if you have some time to upload them.

I'm really not too familiar with the whole setup -- or the way things work. I've seen that the infotainment box (The radio with a screen), or what I believe you're more accurately calling the LG LTE/Nav/etc box has an i.MX chip. I've always thought the vertical chip encased in aluminum on the infotainment board was the Sirius component. I'll have to read into it one of these days. Picked up an older Sirius radio with lifetime XM, would love to flash it's credentials onto my daily driver (although I'm sure it's read-only)

Think I've got enough to update the system via usb (only with the software it currently has), but IDK. Got in my car a few days ago and the screen was black (with backlight still lit) for like 3 minutes or so. And this is on my virgin, untouched, factory infotainment system. Thought it was broken or had a loose connection or something. Eventually kicked back on. Maybe it was updating, but it's hard to tell. AFAIK, over the air updates require an internet connection (I've never connected this system to wifi anywhere near here), or LTE. I had a LTE on this box as a promo, but it expired ~2 years ago. I'd like to know what it was doing. The blinkers still made sounds, and I know for a fact that they don't if the infotainment system is disconnected.

IIRC, I was last trying to get shell scripts to boot off the USB. The idea being that I could do something simple like `echo test > dev/usb/file.txt` to be sure that it works, then perhaps get some executables running. My flashing rig looks like it got bent a little out of shape after a recent move. Hoping it didn't rip up any pads.

I can dig into it myself, but google results seem to be clouded with BS. What's the importance of the HMI? I suppose the infotainment system is one form of human-machine interface. Maybe the steering wheel, gas, brake, shifter, etc are connected to this interface?

**Comment by ilikenwf** [comment]
I'll try and remember to post those images at some point, if I forget and you gain interest just poke me about it...

Yes, internet connection is required - even with the modem removed you can connect to wifi and do an OTA update.

I'm not sure on the components, there may be some mix/match going on. The really interesting part is that they talk to each other over the CAN bus, so we could theoretically do something there through sniffing and spoofing too.

I reconnected my modem and got the compass working again, and pushed the onstar button and instructed them to DISCONNECT me entirely. The light on my mirror now no longer lights up and the online only parts of my screen are greyed out unless I connect to wifi. I wonder if it's placebo or if my modem is really not connecting anymore...

I did find that I only need one antenna connected for GPS locking.

**Comment by ilikenwf** [comment]
...also in that recent article they did on Chevy's infotainment privacy issues, I found this board by BERLA for forensics to be interesting:

https://www.washingtonpost.com/technology/2019/12/17/what-does-your-car-know-about-you-we-hacked-chevy-find-out/

![berla board mylink](https://user-images.githubusercontent.com/414778/72213791-08ef1080-34ed-11ea-85d6-db0adeae3625.jpeg)

**Comment by MatthewTingum** [comment]
Huh, interesting. Might have to dump my retail box... because that blank screen incident was not normal.

An easy way to update this thing would be nice. Looks like obliterating it in glue so nothing comes detached is probably the best solution.

So in that picture, the interesting parts to me are the daughter-board or whatever is connected directly to the infotainment board, and the `other board`. The daughter board doesn't seem to be much -- maybe even a pass-through. What's it converting to? Don't recognize that connector. It spits out USB and Ethernet but there's definitely processing before that. I'll have to break out my logic analyzer and monitor the unpopulated header on my board. I'd have to think it serves a similar purpose.

**Comment by MatthewTingum** [comment]
Haven't looked yet, but fcc documents might reveal a little too much.

**Comment by MatthewTingum** [comment]
I switched the repo to private for a while as an unfinished project. Do we know what the "additional hardware" is here? I'm excited about [eMMC interposers](https://www.devicelocksmith.com/2021/09/modding-devices-with-emmc-flash-wip.html). I've got around 20 ready to go pending my BGA soldering skills.
